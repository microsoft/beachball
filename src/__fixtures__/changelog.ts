import fs from 'fs-extra';
import path from 'path';
import _ from 'lodash';
import { SortedChangeTypes } from '../changefile/changeTypes';
import { ChangelogJson } from '../types/ChangeLog';

/** Read the CHANGELOG.md under the given package path, sanitizing any dates for snapshots */
export function readChangelogMd(packagePath: string): string {
  const changelogFile = path.join(packagePath, 'CHANGELOG.md');
  const text = fs.readFileSync(changelogFile, { encoding: 'utf-8' });
  return text.replace(/\w\w\w, \d\d \w\w\w [\d :]+?GMT/gm, '(date)');
}

/** Read the CHANGELOG.json under the given package path */
export function readChangelogJson(packagePath: string, cleanForSnapshot: boolean = false): ChangelogJson {
  const changelogJsonFile = path.join(packagePath, 'CHANGELOG.json');
  const json = fs.readJSONSync(changelogJsonFile, { encoding: 'utf-8' });
  return cleanForSnapshot ? cleanChangelogJson(json) : json;
}

/**
 * Clean changelog json for a snapshot: replace dates and SHAs with placeholders.
 * Note: this clones the changelog object rather than modifying the original.
 */
export function cleanChangelogJson(changelog: ChangelogJson): ChangelogJson {
  changelog = _.cloneDeep(changelog);
  // for a better snapshot, make the fake commit match if the real commit did
  const fakeCommits: { [commit: string]: string } = {};
  let fakeHashNum = 0;

  for (const entry of changelog.entries) {
    entry.date = '(date)';
    for (const changeType of SortedChangeTypes) {
      if (entry.comments[changeType]) {
        for (const comment of entry.comments[changeType]!) {
          if (!fakeCommits[comment.commit]) {
            fakeCommits[comment.commit] = `(sha1-${fakeHashNum++})`;
          }
          comment.commit = fakeCommits[comment.commit];
        }
      }
    }
  }
  return changelog;
}
