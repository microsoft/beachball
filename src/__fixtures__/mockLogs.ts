import { jest, afterEach, beforeAll, afterAll } from '@jest/globals';
import type { SpyInstance } from 'jest-mock';

/** Methods that will be mocked. More could be added later if needed. */
type MockLogMethod = 'log' | 'warn' | 'error';
const mockedMethods: MockLogMethod[] = ['log', 'warn', 'error'];

type MockLogsOptions = {
  /**
   * Whether to also log to the real console (or subset of methods to log to the console).
   * All logging can be enabled by setting the VERBOSE env var.
   */
  alsoLog?: boolean | MockLogMethod[];
};

export type MockLogs = {
  /** Mocked methods (to access calls etc) */
  mocks: { [k in MockLogMethod]: SpyInstance<typeof console.log> };

  /** Set override options for one test only */
  setOverrideOptions: (options: MockLogsOptions) => void;

  /** Call `mockClear` on all mocks to reset calls */
  clear: () => void;

  /** Get the lines logged to a particular method (or all methods) */
  getMockLines: (method: MockLogMethod | 'all', sanitizeGuids?: boolean) => string;
};

/**
 * Initialize console log mocks, which will be reset after each test. This should be called **outside**
 * of any lifecycle hooks or tests because it calls lifecycle hooks internally for setup and teardown.
 */
export function initMockLogs(options: MockLogsOptions = {}): MockLogs {
  const { alsoLog } = options;
  let allLines: unknown[][] = [];
  let overrideOptions: MockLogsOptions | undefined;
  const jestConsole = { ...console };

  const logs: MockLogs = {
    mocks: {} as MockLogs['mocks'],
    setOverrideOptions: options => {
      overrideOptions = options;
    },
    clear: () => {
      allLines = [];
      overrideOptions = undefined;
      Object.values(logs.mocks).forEach(mock => mock.mockClear());
    },
    getMockLines: (method, sanitizeGuids) => {
      const lines = (method === 'all' ? allLines : logs.mocks[method].mock.calls)
        .map(args =>
          args
            .join(' ')
            .split('\n')
            .map(l => l.trimEnd()) // prevent trailing whitespace in snapshots
            .join('\n')
        )
        .join('\n')
        .trim();

      return sanitizeGuids
        ? // replace change file name GUIDs
          lines.replace(/[a-f\d]{8}(-[a-f\d]{4}){3}-[a-f\d]{12}/g, '<guid>')
        : lines;
    },
  };

  beforeAll(() => {
    for (const method of mockedMethods) {
      const mainShouldLog = shouldLog(method, alsoLog);

      logs.mocks[method] = jest.spyOn(console, method).mockImplementation((...args) => {
        const currentShouldLog =
          overrideOptions === undefined ? mainShouldLog : shouldLog(method, overrideOptions.alsoLog);
        allLines.push([`[${method}]`, ...args]);
        if (process.env.VERBOSE || currentShouldLog) {
          jestConsole[method](...args);
        }
      });
    }
  });

  afterEach(() => {
    logs.clear();
  });

  afterAll(() => {
    Object.values(logs.mocks).forEach(mock => mock.mockRestore());
  });

  return logs;
}

function shouldLog(method: MockLogMethod, alsoLog: boolean | MockLogMethod[] | undefined) {
  return typeof alsoLog === 'boolean' ? alsoLog : alsoLog?.includes(method);
}
