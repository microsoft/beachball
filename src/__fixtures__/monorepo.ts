import path from 'path';
import * as fs from 'fs-extra';
import { BeachballOptions } from '../types/BeachballOptions';
import { BaseRepositoryFactory, Repository } from './repository';
import { PackageJson } from '../types/PackageInfo';

export const packageJsonFixtures: { [path: string]: PackageJson } = {
  'packages/foo': {
    name: 'foo',
    version: '1.0.0',
    dependencies: {
      bar: '^1.3.4',
    },
    main: 'src/index.ts',
    onPublish: {
      main: 'lib/index.js',
    },
    afterPublish: {
      notify: 'message',
    },
  } as PackageJson,

  'packages/bar': {
    name: 'bar',
    version: '1.3.4',
    dependencies: {
      baz: '^1.3.4',
    },
  },

  'packages/baz': {
    name: 'baz',
    version: '1.3.4',
  },

  'packages/grouped/a': {
    name: 'a',
    version: '3.1.2',
  },

  'packages/grouped/b': {
    name: 'b',
    version: '3.1.2',
    dependencies: ['bar'] as any,
  },
};

const beachballConfigFixture = {
  groups: [
    {
      disallowedChangeTypes: null,
      name: 'grouped',
      include: 'grouped*',
    },
  ],
} as BeachballOptions;

export class MonoRepoFactory extends BaseRepositoryFactory {
  constructor() {
    super('beachball-monorepo-upstream-');
  }

  protected initFixture(tmpRepo: Repository) {
    for (const pkg of Object.keys(packageJsonFixtures)) {
      const packageJsonFixture = packageJsonFixtures[pkg];
      const packageJsonFile = path.join(pkg, 'package.json');

      fs.mkdirpSync(path.join(tmpRepo.rootPath, pkg));

      fs.writeJSONSync(path.join(tmpRepo.rootPath, packageJsonFile), packageJsonFixture, {
        spaces: 2,
      });
      tmpRepo.commitChange(packageJsonFile);
    }

    tmpRepo.commitChange('package.json', JSON.stringify({ name: 'monorepo-fixture', version: '1.0.0' }, null, 2));

    tmpRepo.commitChange('beachball.config.js', 'module.exports = ' + JSON.stringify(beachballConfigFixture, null, 2));
  }
}
