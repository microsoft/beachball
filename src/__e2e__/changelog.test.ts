import path from 'path';
import fs from 'fs-extra';
import _ from 'lodash';

import { initMockLogs } from '../__fixtures__/mockLogs';
import { MonoRepoFactory, packageJsonFixtures } from '../__fixtures__/monorepo';
import { RepositoryFactory } from '../__fixtures__/repository';

import { writeChangelog } from '../changelog/writeChangelog';
import { getPackageInfos } from '../monorepo/getPackageInfos';
import { writeChangeFiles } from '../changefile/writeChangeFiles';
import { readChangeFiles } from '../changefile/readChangeFiles';
import { SortedChangeTypes } from '../changefile/getPackageChangeTypes';
import { BeachballOptions } from '../types/BeachballOptions';
import { ChangeFileInfo, ChangeInfo } from '../types/ChangeInfo';
import { ChangelogJson } from '../types/ChangeLog';

function getChange(packageName: string, comment: string): ChangeFileInfo {
  return {
    comment,
    email: 'test@testtestme.com',
    packageName,
    type: 'patch',
    dependentChangeType: 'patch',
  };
}

function cleanMarkdownForSnapshot(text: string) {
  return text.replace(/\w\w\w, \d\d \w\w\w [\d :]+?GMT/gm, '(date)');
}

function cleanJsonForSnapshot(changelog: ChangelogJson) {
  changelog = _.cloneDeep(changelog);
  // for a better snapshot, make the fake commit match if the real commit did
  const fakeCommits: { [commit: string]: string } = {};
  let fakeHashNum = 0;

  for (const entry of changelog.entries) {
    entry.date = '(date)';
    for (const changeType of SortedChangeTypes) {
      if (entry.comments[changeType]) {
        for (const comment of entry.comments[changeType]!) {
          if (!fakeCommits[comment.commit]) {
            fakeCommits[comment.commit] = `(sha1-${fakeHashNum++})`;
          }
          comment.commit = fakeCommits[comment.commit];
        }
      }
    }
  }
  return changelog;
}

describe('changelog generation', () => {
  let repositoryFactory: RepositoryFactory;
  let monoRepoFactory: MonoRepoFactory;

  const logs = initMockLogs();

  beforeAll(() => {
    repositoryFactory = new RepositoryFactory();
    repositoryFactory.create();
    monoRepoFactory = new MonoRepoFactory();
    monoRepoFactory.create();
  });

  afterAll(() => {
    repositoryFactory.cleanUp();
    monoRepoFactory.cleanUp();
  });

  describe('readChangeFiles', () => {
    it('does not add commit hash', () => {
      const repository = repositoryFactory.cloneRepository();
      repository.commitChange('foo');
      writeChangeFiles({ changes: [getChange('foo', 'comment 1')], cwd: repository.rootPath });

      const packageInfos = getPackageInfos(repository.rootPath);
      const changeSet = readChangeFiles({ path: repository.rootPath } as BeachballOptions, packageInfos);
      expect(changeSet).toHaveLength(1);
      expect(changeSet[0].change.commit).toBe(undefined);
    });

    it('excludes invalid change files', () => {
      const monoRepo = monoRepoFactory.cloneRepository();
      monoRepo.commitChange(
        'packages/bar/package.json',
        JSON.stringify({ ...packageJsonFixtures['packages/bar'], private: true })
      );
      // fake doesn't exist, bar is private, foo is okay
      writeChangeFiles({
        changes: [getChange('fake', 'comment 1'), getChange('bar', 'comment 2'), getChange('foo', 'comment 3')],
        cwd: monoRepo.rootPath,
      });

      const packageInfos = getPackageInfos(monoRepo.rootPath);
      const changeSet = readChangeFiles({ path: monoRepo.rootPath } as BeachballOptions, packageInfos);
      expect(changeSet).toHaveLength(1);

      expect(logs.mocks.warn).toHaveBeenCalledWith(expect.stringContaining('Change detected for private package bar'));
      expect(logs.mocks.warn).toHaveBeenCalledWith(
        expect.stringContaining('Change detected for nonexistent package fake')
      );
    });

    it('excludes invalid changes from grouped change file', () => {
      const monoRepo = monoRepoFactory.cloneRepository();
      monoRepo.commitChange(
        'packages/bar/package.json',
        JSON.stringify({ ...packageJsonFixtures['packages/bar'], private: true })
      );
      // fake doesn't exist, bar is private, foo is okay
      writeChangeFiles({
        changes: [getChange('fake', 'comment 1'), getChange('bar', 'comment 2'), getChange('foo', 'comment 3')],
        cwd: monoRepo.rootPath,
        groupChanges: true,
      });

      const packageInfos = getPackageInfos(monoRepo.rootPath);
      const changeSet = readChangeFiles(
        { path: monoRepo.rootPath, groupChanges: true } as BeachballOptions,
        packageInfos
      );
      expect(changeSet).toHaveLength(1);

      expect(logs.mocks.warn).toHaveBeenCalledWith(expect.stringContaining('Change detected for private package bar'));
      expect(logs.mocks.warn).toHaveBeenCalledWith(
        expect.stringContaining('Change detected for nonexistent package fake')
      );
    });

    it('excludes out of scope change files', () => {
      const monoRepo = monoRepoFactory.cloneRepository();
      writeChangeFiles({
        changes: [getChange('bar', 'comment 2'), getChange('foo', 'comment 3')],
        cwd: monoRepo.rootPath,
      });

      const packageInfos = getPackageInfos(monoRepo.rootPath);
      const changeSet = readChangeFiles(
        { path: monoRepo.rootPath, scope: ['packages/foo'] } as BeachballOptions,
        packageInfos
      );
      expect(changeSet).toHaveLength(1);
      expect(logs.mocks.warn).not.toHaveBeenCalled();
    });

    it('excludes out of scope changes from grouped change file', () => {
      const monoRepo = monoRepoFactory.cloneRepository();
      writeChangeFiles({
        changes: [getChange('bar', 'comment 2'), getChange('foo', 'comment 3')],
        cwd: monoRepo.rootPath,
        groupChanges: true,
      });

      const packageInfos = getPackageInfos(monoRepo.rootPath);
      const changeSet = readChangeFiles(
        { path: monoRepo.rootPath, scope: ['packages/foo'], groupChanges: true } as BeachballOptions,
        packageInfos
      );
      expect(changeSet).toHaveLength(1);
      expect(logs.mocks.warn).not.toHaveBeenCalled();
    });
  });

  describe('writeChangelog', () => {
    it('generates correct changelog', async () => {
      const repository = repositoryFactory.cloneRepository();
      repository.commitChange('foo');
      writeChangeFiles({ changes: [getChange('foo', 'additional comment 2')], cwd: repository.rootPath });
      writeChangeFiles({ changes: [getChange('foo', 'additional comment 1')], cwd: repository.rootPath });
      writeChangeFiles({ changes: [getChange('foo', 'comment 1')], cwd: repository.rootPath });

      repository.commitChange('bar');
      writeChangeFiles({ changes: [getChange('foo', 'comment 2')], cwd: repository.rootPath });

      const beachballOptions = { path: repository.rootPath } as BeachballOptions;
      const packageInfos = getPackageInfos(repository.rootPath);
      const changes = readChangeFiles(beachballOptions, packageInfos);

      await writeChangelog(beachballOptions, changes, { foo: 'patch' }, { foo: new Set(['foo']) }, packageInfos);

      const changelogFile = path.join(repository.rootPath, 'CHANGELOG.md');
      const text = fs.readFileSync(changelogFile, { encoding: 'utf-8' });
      expect(cleanMarkdownForSnapshot(text)).toMatchSnapshot();

      const changelogJsonFile = path.join(repository.rootPath, 'CHANGELOG.json');
      const changelogJson: ChangelogJson = fs.readJSONSync(changelogJsonFile);
      expect(cleanJsonForSnapshot(changelogJson)).toMatchSnapshot();

      // Every entry should have a different commit hash
      const patchComments = changelogJson.entries[0].comments.patch!;
      const commits = patchComments.map(entry => entry.commit);
      expect(new Set(commits).size).toEqual(patchComments.length);

      // The first entry should be the newest
      expect(patchComments[0].commit).toBe(repository.getCurrentHash());
    });

    it('generates correct changelog in monorepo with groupChanges (grouped change FILES)', async () => {
      const monoRepo = monoRepoFactory.cloneRepository();
      monoRepo.commitChange('foo');
      const options: Omit<Parameters<typeof writeChangeFiles>[0], 'changes'> = {
        cwd: monoRepo.rootPath,
        groupChanges: true,
      };
      writeChangeFiles({
        changes: [getChange('foo', 'additional comment 2'), getChange('bar', 'comment from bar change ')],
        ...options,
      });
      writeChangeFiles({ changes: [getChange('foo', 'additional comment 1')], ...options });
      writeChangeFiles({ changes: [getChange('foo', 'comment 1')], ...options });

      monoRepo.commitChange('bar');
      writeChangeFiles({ changes: [getChange('foo', 'comment 2')], ...options });

      const beachballOptions = { path: monoRepo.rootPath, groupChanges: true } as BeachballOptions;
      const packageInfos = getPackageInfos(monoRepo.rootPath);
      const changes = readChangeFiles(beachballOptions, packageInfos);

      await writeChangelog(beachballOptions, changes, { foo: 'patch', bar: 'patch' }, {}, packageInfos);

      // check changelogs for both foo and bar
      const fooText = fs.readFileSync(path.join(monoRepo.rootPath, 'packages/foo/CHANGELOG.md'), { encoding: 'utf-8' });
      expect(cleanMarkdownForSnapshot(fooText)).toMatchSnapshot();
      const barText = fs.readFileSync(path.join(monoRepo.rootPath, 'packages/bar/CHANGELOG.md'), { encoding: 'utf-8' });
      expect(cleanMarkdownForSnapshot(barText)).toMatchSnapshot();

      const fooJson: ChangelogJson = fs.readJSONSync(path.join(monoRepo.rootPath, 'packages/foo/CHANGELOG.json'));
      expect(cleanJsonForSnapshot(fooJson)).toMatchSnapshot();
      const barJson: ChangelogJson = fs.readJSONSync(path.join(monoRepo.rootPath, 'packages/bar/CHANGELOG.json'));
      expect(cleanJsonForSnapshot(barJson)).toMatchSnapshot();

      // Every entry should have a different commit hash
      const patchComments = fooJson.entries[0].comments.patch!;
      const commits = patchComments.map(entry => entry.commit);
      expect(new Set(commits).size).toEqual(patchComments.length);

      // The first entry should be the newest
      expect(patchComments[0].commit).toBe(monoRepo.getCurrentHash());
    });

    it('generates correct grouped changelog', async () => {
      const monoRepo = monoRepoFactory.cloneRepository();
      monoRepo.commitChange('foo');
      writeChangeFiles({ changes: [getChange('foo', 'comment 1')], cwd: monoRepo.rootPath });

      monoRepo.commitChange('bar');
      writeChangeFiles({ changes: [getChange('bar', 'comment 2')], cwd: monoRepo.rootPath });

      writeChangeFiles({ changes: [getChange('bar', 'comment 3')], cwd: monoRepo.rootPath });

      const beachballOptions = {
        path: monoRepo.rootPath,
        changelog: {
          groups: [
            {
              masterPackageName: 'foo',
              changelogPath: monoRepo.rootPath,
              include: ['packages/foo', 'packages/bar'],
            },
          ],
        },
      } as BeachballOptions;

      const packageInfos = getPackageInfos(monoRepo.rootPath);
      const changes = readChangeFiles(beachballOptions, packageInfos);

      await writeChangelog(beachballOptions, changes, {}, {}, packageInfos);

      // Validate changelog for foo package
      const fooChangelogFile = path.join(monoRepo.rootPath, 'packages', 'foo', 'CHANGELOG.md');
      const fooChangelogText = fs.readFileSync(fooChangelogFile, { encoding: 'utf-8' });
      expect(cleanMarkdownForSnapshot(fooChangelogText)).toMatchSnapshot();

      // Validate changelog for bar package
      const barChangelogFile = path.join(monoRepo.rootPath, 'packages', 'bar', 'CHANGELOG.md');
      const barChangelogText = fs.readFileSync(barChangelogFile, { encoding: 'utf-8' });
      expect(cleanMarkdownForSnapshot(barChangelogText)).toMatchSnapshot();

      // Validate grouped changelog for foo and bar packages
      const groupedChangelogFile = path.join(monoRepo.rootPath, 'CHANGELOG.md');
      const groupedChangelogText = fs.readFileSync(groupedChangelogFile, { encoding: 'utf-8' });
      expect(cleanMarkdownForSnapshot(groupedChangelogText)).toMatchSnapshot();
    });

    it('generates grouped changelog without dependent change entries', async () => {
      const monoRepo = monoRepoFactory.cloneRepository();
      monoRepo.commitChange('baz');
      writeChangeFiles({ changes: [getChange('baz', 'comment 1')], cwd: monoRepo.rootPath });

      const beachballOptions = {
        path: monoRepo.rootPath,
        changelog: {
          groups: [
            {
              masterPackageName: 'foo',
              changelogPath: monoRepo.rootPath,
              include: ['packages/foo', 'packages/bar', 'packages/baz'],
            },
          ],
        },
      } as BeachballOptions;

      const packageInfos = getPackageInfos(monoRepo.rootPath);
      const changes = readChangeFiles(beachballOptions, packageInfos);

      await writeChangelog(
        beachballOptions,
        changes,
        { bar: 'patch', baz: 'patch' },
        { bar: new Set(['baz']) },
        packageInfos
      );

      // Validate changelog for bar package
      const barChangelogFile = path.join(monoRepo.rootPath, 'packages', 'bar', 'CHANGELOG.md');
      const barChangelogText = fs.readFileSync(barChangelogFile, { encoding: 'utf-8' });
      expect(barChangelogText).toContain('- Bump baz');
      expect(cleanMarkdownForSnapshot(barChangelogText)).toMatchSnapshot();

      // Validate changelog for baz package
      const bazChangelogFile = path.join(monoRepo.rootPath, 'packages', 'baz', 'CHANGELOG.md');
      const bazChangelogText = fs.readFileSync(bazChangelogFile, { encoding: 'utf-8' });
      expect(cleanMarkdownForSnapshot(bazChangelogText)).toMatchSnapshot();

      // Validate grouped changelog for foo master package
      const groupedChangelogFile = path.join(monoRepo.rootPath, 'CHANGELOG.md');
      const groupedChangelogText = fs.readFileSync(groupedChangelogFile, { encoding: 'utf-8' });
      expect(groupedChangelogText).toContain('- comment 1');
      expect(groupedChangelogText).not.toContain('- Bump baz');
      expect(cleanMarkdownForSnapshot(groupedChangelogText)).toMatchSnapshot();
    });

    it('generates grouped changelog without dependent change entries where packages have normal changes and dependency changes', async () => {
      const monoRepo = monoRepoFactory.cloneRepository();
      monoRepo.commitChange('baz');
      writeChangeFiles({ changes: [getChange('baz', 'comment 1')], cwd: monoRepo.rootPath });
      writeChangeFiles({ changes: [getChange('bar', 'comment 1')], cwd: monoRepo.rootPath });

      const beachballOptions = {
        path: monoRepo.rootPath,
        changelog: {
          groups: [
            {
              masterPackageName: 'foo',
              changelogPath: monoRepo.rootPath,
              include: ['packages/foo', 'packages/bar', 'packages/baz'],
            },
          ],
        },
      } as BeachballOptions;

      const packageInfos = getPackageInfos(monoRepo.rootPath);
      const changes = readChangeFiles(beachballOptions, packageInfos);

      await writeChangelog(
        beachballOptions,
        changes,
        { bar: 'patch', baz: 'patch' },
        { bar: new Set(['baz']) },
        packageInfos
      );

      // Validate changelog for bar package
      const barChangelogFile = path.join(monoRepo.rootPath, 'packages', 'bar', 'CHANGELOG.md');
      const barChangelogText = fs.readFileSync(barChangelogFile, { encoding: 'utf-8' });
      expect(cleanMarkdownForSnapshot(barChangelogText)).toMatchSnapshot();

      // Validate changelog for baz package
      const bazChangelogFile = path.join(monoRepo.rootPath, 'packages', 'baz', 'CHANGELOG.md');
      const bazChangelogText = fs.readFileSync(bazChangelogFile, { encoding: 'utf-8' });
      expect(cleanMarkdownForSnapshot(bazChangelogText)).toMatchSnapshot();

      // Validate grouped changelog for foo master package
      const groupedChangelogFile = path.join(monoRepo.rootPath, 'CHANGELOG.md');
      const groupedChangelogText = fs.readFileSync(groupedChangelogFile, { encoding: 'utf-8' });
      expect(cleanMarkdownForSnapshot(groupedChangelogText)).toMatchSnapshot();
    });

    it('generates correct grouped changelog when grouped change log is saved to the same dir as a regular changelog', async () => {
      const monoRepo = monoRepoFactory.cloneRepository();
      monoRepo.commitChange('foo');
      writeChangeFiles({ changes: [getChange('foo', 'comment 1')], cwd: monoRepo.rootPath });

      monoRepo.commitChange('bar');
      writeChangeFiles({ changes: [getChange('bar', 'comment 2')], cwd: monoRepo.rootPath });

      const beachballOptions = {
        path: monoRepo.rootPath,
        changelog: {
          groups: [
            {
              masterPackageName: 'foo',
              changelogPath: path.join(monoRepo.rootPath, 'packages', 'foo'),
              include: ['packages/foo', 'packages/bar'],
            },
          ],
        },
      } as BeachballOptions;

      const packageInfos = getPackageInfos(monoRepo.rootPath);
      const changes = readChangeFiles(beachballOptions, packageInfos);

      await writeChangelog(beachballOptions, changes, {}, {}, packageInfos);

      // Validate changelog for bar package
      const barChangelogFile = path.join(monoRepo.rootPath, 'packages', 'bar', 'CHANGELOG.md');
      const barChangelogText = fs.readFileSync(barChangelogFile, { encoding: 'utf-8' });
      expect(cleanMarkdownForSnapshot(barChangelogText)).toMatchSnapshot();

      // Validate grouped changelog for foo and bar packages
      const groupedChangelogFile = path.join(monoRepo.rootPath, 'packages', 'foo', 'CHANGELOG.md');
      const groupedChangelogText = fs.readFileSync(groupedChangelogFile, { encoding: 'utf-8' });
      expect(cleanMarkdownForSnapshot(groupedChangelogText)).toMatchSnapshot();
    });

    it('Verify that the changeFile transform functions are run, if provided', async () => {
      const editedComment: string = 'Edited comment for testing';
      const monoRepo = monoRepoFactory.cloneRepository();
      monoRepo.commitChange('foo');
      writeChangeFiles({ changes: [getChange('foo', 'comment 1')], cwd: monoRepo.rootPath });

      monoRepo.commitChange('bar');
      writeChangeFiles({ changes: [getChange('bar', 'comment 2')], cwd: monoRepo.rootPath });

      const beachballOptions = {
        path: monoRepo.rootPath,
        transform: {
          changeFiles: (changeFile: ChangeInfo, changeFilePath) => {
            // For test, we will be changing the comment based on the package name
            if (changeFile.packageName === 'foo') {
              changeFile.comment = editedComment;
            }
            return changeFile;
          },
        },
        changelog: {
          groups: [
            {
              masterPackageName: 'foo',
              changelogPath: path.join(monoRepo.rootPath, 'packages', 'foo'),
              include: ['packages/foo', 'packages/bar'],
            },
          ],
        },
      } as BeachballOptions;

      const packageInfos = getPackageInfos(monoRepo.rootPath);
      const changes = readChangeFiles(beachballOptions, packageInfos);

      // Verify that the comment of only the intended change file is changed
      for (const { change, changeFile } of changes) {
        if (changeFile.substr(0, 3) === 'foo') {
          expect(change.comment).toBe(editedComment);
        } else {
          expect(change.comment).toBe('comment 2');
        }
      }
    });
  });
});
