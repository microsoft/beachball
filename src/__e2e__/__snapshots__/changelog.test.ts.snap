// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`changelog generation writeChangelog generates correct changelog 1`] = `
"# Change Log - foo

This log was last generated on (date) and should not be manually modified.

<!-- Start content -->

## 1.0.0

(date)

### Patches

- comment 2 (test@testtestme.com)
- comment 1 (test@testtestme.com)
- additional comment 1 (test@testtestme.com)
- additional comment 2 (test@testtestme.com)
"
`;

exports[`changelog generation writeChangelog generates correct changelog 2`] = `
Object {
  "entries": Array [
    Object {
      "comments": Object {
        "patch": Array [
          Object {
            "author": "test@testtestme.com",
            "comment": "comment 2",
            "commit": "(sha1-0)",
            "package": "foo",
          },
          Object {
            "author": "test@testtestme.com",
            "comment": "comment 1",
            "commit": "(sha1-1)",
            "package": "foo",
          },
          Object {
            "author": "test@testtestme.com",
            "comment": "additional comment 1",
            "commit": "(sha1-2)",
            "package": "foo",
          },
          Object {
            "author": "test@testtestme.com",
            "comment": "additional comment 2",
            "commit": "(sha1-3)",
            "package": "foo",
          },
        ],
      },
      "date": "(date)",
      "tag": "foo_v1.0.0",
      "version": "1.0.0",
    },
  ],
  "name": "foo",
}
`;

exports[`changelog generation writeChangelog generates correct changelog in monorepo with groupChanges (grouped change FILES) 1`] = `
"# Change Log - foo

This log was last generated on (date) and should not be manually modified.

<!-- Start content -->

## 1.0.0

(date)

### Patches

- comment 2 (test@testtestme.com)
- comment 1 (test@testtestme.com)
- additional comment 1 (test@testtestme.com)
- additional comment 2 (test@testtestme.com)
"
`;

exports[`changelog generation writeChangelog generates correct changelog in monorepo with groupChanges (grouped change FILES) 2`] = `
"# Change Log - bar

This log was last generated on (date) and should not be manually modified.

<!-- Start content -->

## 1.3.4

(date)

### Patches

- comment from bar change  (test@testtestme.com)
"
`;

exports[`changelog generation writeChangelog generates correct changelog in monorepo with groupChanges (grouped change FILES) 3`] = `
Object {
  "entries": Array [
    Object {
      "comments": Object {
        "patch": Array [
          Object {
            "author": "test@testtestme.com",
            "comment": "comment 2",
            "commit": "(sha1-0)",
            "package": "foo",
          },
          Object {
            "author": "test@testtestme.com",
            "comment": "comment 1",
            "commit": "(sha1-1)",
            "package": "foo",
          },
          Object {
            "author": "test@testtestme.com",
            "comment": "additional comment 1",
            "commit": "(sha1-2)",
            "package": "foo",
          },
          Object {
            "author": "test@testtestme.com",
            "comment": "additional comment 2",
            "commit": "(sha1-3)",
            "package": "foo",
          },
        ],
      },
      "date": "(date)",
      "tag": "foo_v1.0.0",
      "version": "1.0.0",
    },
  ],
  "name": "foo",
}
`;

exports[`changelog generation writeChangelog generates correct changelog in monorepo with groupChanges (grouped change FILES) 4`] = `
Object {
  "entries": Array [
    Object {
      "comments": Object {
        "patch": Array [
          Object {
            "author": "test@testtestme.com",
            "comment": "comment from bar change ",
            "commit": "(sha1-0)",
            "package": "bar",
          },
        ],
      },
      "date": "(date)",
      "tag": "bar_v1.3.4",
      "version": "1.3.4",
    },
  ],
  "name": "bar",
}
`;

exports[`changelog generation writeChangelog generates correct grouped changelog 1`] = `
"# Change Log - foo

This log was last generated on (date) and should not be manually modified.

<!-- Start content -->

## 1.0.0

(date)

### Patches

- comment 1 (test@testtestme.com)
"
`;

exports[`changelog generation writeChangelog generates correct grouped changelog 2`] = `
"# Change Log - bar

This log was last generated on (date) and should not be manually modified.

<!-- Start content -->

## 1.3.4

(date)

### Patches

- comment 3 (test@testtestme.com)
- comment 2 (test@testtestme.com)
"
`;

exports[`changelog generation writeChangelog generates correct grouped changelog 3`] = `
"# Change Log - foo

This log was last generated on (date) and should not be manually modified.

<!-- Start content -->

## 1.0.0

(date)

### Patches

- \`bar\`
  - comment 3 (test@testtestme.com)
  - comment 2 (test@testtestme.com)
- \`foo\`
  - comment 1 (test@testtestme.com)
"
`;

exports[`changelog generation writeChangelog generates correct grouped changelog when grouped change log is saved to the same dir as a regular changelog 1`] = `
"# Change Log - bar

This log was last generated on (date) and should not be manually modified.

<!-- Start content -->

## 1.3.4

(date)

### Patches

- comment 2 (test@testtestme.com)
"
`;

exports[`changelog generation writeChangelog generates correct grouped changelog when grouped change log is saved to the same dir as a regular changelog 2`] = `
"# Change Log - foo

This log was last generated on (date) and should not be manually modified.

<!-- Start content -->

## 1.0.0

(date)

### Patches

- \`bar\`
  - comment 2 (test@testtestme.com)
- \`foo\`
  - comment 1 (test@testtestme.com)
"
`;

exports[`changelog generation writeChangelog generates grouped changelog without dependent change entries 1`] = `
"# Change Log - bar

This log was last generated on (date) and should not be manually modified.

<!-- Start content -->

## 1.3.4

(date)

### Patches

- Bump baz to v1.3.4
"
`;

exports[`changelog generation writeChangelog generates grouped changelog without dependent change entries 2`] = `
"# Change Log - baz

This log was last generated on (date) and should not be manually modified.

<!-- Start content -->

## 1.3.4

(date)

### Patches

- comment 1 (test@testtestme.com)
"
`;

exports[`changelog generation writeChangelog generates grouped changelog without dependent change entries 3`] = `
"# Change Log - foo

This log was last generated on (date) and should not be manually modified.

<!-- Start content -->

## 1.0.0

(date)

### Patches

- \`baz\`
  - comment 1 (test@testtestme.com)
"
`;

exports[`changelog generation writeChangelog generates grouped changelog without dependent change entries where packages have normal changes and dependency changes 1`] = `
"# Change Log - bar

This log was last generated on (date) and should not be manually modified.

<!-- Start content -->

## 1.3.4

(date)

### Patches

- comment 1 (test@testtestme.com)
- Bump baz to v1.3.4
"
`;

exports[`changelog generation writeChangelog generates grouped changelog without dependent change entries where packages have normal changes and dependency changes 2`] = `
"# Change Log - baz

This log was last generated on (date) and should not be manually modified.

<!-- Start content -->

## 1.3.4

(date)

### Patches

- comment 1 (test@testtestme.com)
"
`;

exports[`changelog generation writeChangelog generates grouped changelog without dependent change entries where packages have normal changes and dependency changes 3`] = `
"# Change Log - foo

This log was last generated on (date) and should not be manually modified.

<!-- Start content -->

## 1.0.0

(date)

### Patches

- \`bar\`
  - comment 1 (test@testtestme.com)
- \`baz\`
  - comment 1 (test@testtestme.com)
"
`;
